### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by maiksaray**
Oh my... 
Ok, here it goes. Threw a bunch of deps into the mix. 
scalatest, scalatest+play, scalatest gradle plugin...

Started playing with scalatest and it turned out that play can't be tested with asyncSpecs, so had to switch to regular specs for a while. 

Craeted whole UserService and adjusted a couple of things on the way

### CODE REVIEW

File: app/controllers/IndexController.scala [hunk 1]
```diff
@@ -4,7 +4,6 @@ import common._
 import common.Serialization.{INSTANCE => Json}
 import dao.{SessionDao, UserDao}
 import javax.inject.{Inject, Singleton}
-import play.api.Logging
```
Comment (maiksaray) [inline app/controllers/IndexController.scala:7]: Logging is imported transiently by SecuredController that we inherit from.

File: app/dao/UserDao.scala [hunk 2]
```diff
@@ -14,20 +14,25 @@ class UserDao @Inject()(repo: UserRepo)(implicit ec: ExecutionContext)
     with Logging {
 
   override def ensureExists(): Future[Unit] = {
-    for {
-      _ <- super.ensureExists()
-      user <- repo.findByName("admin")
-      if user.isEmpty
```
Comment (maiksaray) [inline app/dao/UserDao.scala:20]: This was throwing on else since Future was failed. 
Reworked into more elegant approach

File: .travis.yml [hunk 3]
```diff
@@ -14,4 +14,4 @@ cache:
     - $HOME/.gradle/wrapper/
 
 script:
-  - ./gradlew build
+  - ./gradlew build test
```
Comment (AnastasiiaTitova) [inline .travis.yml:17]: AFAIK tests are run without specifying it...
At least in my project it works like that 100%
Please, check that this change is needed

File: app/controllers/UserController.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  def create(): Action[AnyContent] = securedAsync(Admin.INSTANCE :: Nil, Action.async {
+    request => {
+      //      todo: extract this json processing/bad-requesting
+      val body = request.body.asText
+      body match {
+        case Some(userData) =>
+          val user = Json.fromJson(userData, classOf[common.User])
+          userService.createUser(user).map {
+            user => Ok(Json.toJson(user))
```
Comment (AnastasiiaTitova) [inline app/controllers/UserController.scala:49]: Wut
You're deserializing and then serializing again?

File: app/dao/UserDao.scala [hunk 5]
```diff
@@ -55,14 +58,23 @@ class UserDao @Inject()(repo: UserRepo)(implicit ec: ExecutionContext)
   def auth(user: common.User): Future[Option[common.User]] =
     auth(user.getName, user.getPassword)
 
+  def getUser(id: Long): Future[Option[common.User]] = {
+    //    TODO:rework this!
+    logger.info(s"Obtaining user with id $id")
+    repo.find(id).map { f =>
+      f.map { u =>
```
Comment (AnastasiiaTitova) [inline app/dao/UserDao.scala:65]: `f`
`u`
:panic:

File: app/dao/UserDao.scala [hunk 6]
```diff
@@ -55,14 +58,23 @@ class UserDao @Inject()(repo: UserRepo)(implicit ec: ExecutionContext)
...
     //    TODO:rework this!
     logger.info(s"Obtaining user with name $username")
-    repo.findByName(username).map {
-      f =>
-        f.map { u =>
-          userDbToModel(u)
-        }
+    repo.findByName(username).map { f =>
+      f.map { u =>
+        userDbToModel(u)
```
Comment (AnastasiiaTitova) [inline app/dao/UserDao.scala:76]: :panic: again

File: app/misc/ApplicationStart.scala [hunk 7]
```diff
@@ -17,6 +17,7 @@ class ApplicationStart @Inject()(userDao: UserDao,
 
   logger.info("Starting application, preparing db")
   //TODO: look at how the fuck evolutions should be done cause this is fucked up
+//  ALso, here we go around services layer, cuase we can
```
Comment (AnastasiiaTitova) [inline app/misc/ApplicationStart.scala:20]: c u a s e

File: app/services/UserService.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  def update(name: String, change: common.User => common.User): Future[User] = {
+    userDao.getUser(name).map {
+      case None => throw new YaltaBaseException(s"User $name does not exist, can't update")
+      case Some(user) => user
+    }.flatMap {
+      user => userDao.update(change(user))
+    }
+  }
+
+  def changeRole(name: String, newRole: common.Role): Future[common.User] = {
```
Comment (AnastasiiaTitova) [inline app/services/UserService.scala:39]: R we doin this????? Why??

File: test/AsyncPlaySpec.scala [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+import org.scalatestplus.play.WsScalaTestClient
+
+/**
+ * This is custom implementation of PlaySpec, but extending AsyncWordSpec instead of WordSpec
+ * This does not work properly now for number of reasons
+ * https://github.com/playframework/scalatestplus-play/issues/112
+ * Left here for future references
+ */
+abstract class AsyncPlaySpec extends AsyncWordSpec with MustMatchers with OptionValues with WsScalaTestClient{
+
```
Comment (AnastasiiaTitova) [inline test/AsyncPlaySpec.scala:11]: empty line?

File: test/UserServiceSpec.scala [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    "Throw for usernames starting with digit" in {
+      assertThrows[InvalidDataException] {
+        userService.createUser("1", "1", common.Admin.INSTANCE)
+      }
+    }
+  }
+
+}
```
Comment (AnastasiiaTitova) [inline test/UserServiceSpec.scala:29]: newline?

**Review by AnastasiiaTitova — COMMENTED** [review]
Maybe better description?

File: app/controllers/UserController.scala (hunk 4 above)
Comment (maiksaray) [reply app/controllers/UserController.scala:49]: yep, I deserialize content that came in http request body, process it (shove it down `UserService`) and get resulting created user (with id and stuff) and serialize it back to return as http response

File: app/dao/UserDao.scala (hunk 5 above)
Comment (maiksaray) [reply app/dao/UserDao.scala:65]: yeaah, there's `todo` right next to it, not sure how to make it better =(

File: app/services/UserService.scala (hunk 8 above)
Comment (maiksaray) [reply app/services/UserService.scala:39]: Why not?

File: app/controllers/UserController.scala (hunk 4 above)
Comment (AnastasiiaTitova) [reply app/controllers/UserController.scala:49]: Jeez

File: app/services/UserService.scala (hunk 8 above)
Comment (AnastasiiaTitova) [reply app/services/UserService.scala:39]: I strongly believe that admin should be admin forever
And driver should forever be driver
If person wants to be both - let's create two different accounts
Same with role migration

File: app/dao/UserDao.scala (hunk 5 above)
Comment (AnastasiiaTitova) [reply app/dao/UserDao.scala:65]: ANYTHING but this :C

File: test/UserServiceSpec.scala (hunk 10 above)
Comment (maiksaray) [reply test/UserServiceSpec.scala:29]: removed

File: test/AsyncPlaySpec.scala (hunk 9 above)
Comment (maiksaray) [reply test/AsyncPlaySpec.scala]: cleaned

File: app/misc/ApplicationStart.scala (hunk 7 above)
Comment (maiksaray) [reply app/misc/ApplicationStart.scala]: =)

File: app/dao/UserDao.scala (hunk 6 above)
Comment (maiksaray) [reply app/dao/UserDao.scala]: made it more readable

File: .travis.yml (hunk 3 above)
Comment (maiksaray) [reply .travis.yml]: You were right :
gradlew build --dry-run
:minifyPlayJavaScript SKIPPED
:createPlayAssetsJar SKIPPED
:compileKotlin SKIPPED
:compileJava SKIPPED
:compilePlayRoutes SKIPPED
:compilePlayTwirlTemplates SKIPPED
:compileScala SKIPPED
:processResources SKIPPED
:classes SKIPPED
:inspectClassesForKotlinIC SKIPPED
:jar SKIPPED
:createMainDistributionJar SKIPPED
:createMainStartScripts SKIPPED
:distTar SKIPPED
:distZip SKIPPED
:assemble SKIPPED
:compileTestKotlin SKIPPED
:compileTestJava SKIPPED
:compileTestScala SKIPPED
:processTestResources SKIPPED
:testClasses SKIPPED
**:test SKIPPED**
:check SKIPPED
:build SKIPPED

File: app/services/UserService.scala (hunk 8 above)
Comment (maiksaray) [reply app/services/UserService.scala:39]: Hmmm, let's leave it like that for now and just not call this from API, k?

Comment (AnastasiiaTitova) [reply app/services/UserService.scala:39]: Meh
k

**Review by AnastasiiaTitova — APPROVED** [review]

**Pull request merged**